File: Projeto Didio Pizza/Didio Pizza Springboot/backend/src/main/java/com/pizzaria/didiopizza/service/PedidoService.java
package com.pizzaria.didiopizza.service;

import java.util.List;
import java.util.Map;

import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import com.pizzaria.didiopizza.repository.CarrinhoRepository;
import com.pizzaria.didiopizza.repository.EnderecoRepository;
import com.pizzaria.didiopizza.repository.PedidoRepository;

@Service
public class PedidoService {
    private final CarrinhoRepository carrinho;
    private final PedidoRepository pedido;
    private final EnderecoRepository endereco;
    private final LojaService loja;

    public PedidoService(CarrinhoRepository c, PedidoRepository p, EnderecoRepository e, LojaService l) {
        carrinho = c;
        pedido = p;
        endereco = e;
        loja = l;
    }

    @Transactional
    public Integer fecharPedido(Integer clienteId, Integer enderecoId) {
        loja.verificarLojaAberta();

        enderecoId = resolverEndereco(clienteId, enderecoId);

        var car = carrinho.buscarCarrinhoAtivo(clienteId);
        if (car == null)
            throw new RuntimeException("Carrinho vazio.");

        Integer carrinhoId = ((Number) car.get("id")).intValue();

        var itens = carrinho.listarItens(carrinhoId);
        if (itens.isEmpty())
            throw new RuntimeException("Carrinho sem itens.");

        Integer id = criarPedidoComLista(clienteId, enderecoId, itens);

        carrinho.limparCarrinho(carrinhoId);

        return id;
    }

    @Transactional
    public Integer fecharPedidoComItens(Integer clienteId, Integer enderecoId, List<Map<String, Object>> itens) {
        loja.verificarLojaAberta();

        if (itens == null || itens.isEmpty())
            throw new RuntimeException("Pedido sem itens.");

        enderecoId = resolverEndereco(clienteId, enderecoId);

        return criarPedidoComLista(clienteId, enderecoId, itens);
    }

    private Integer resolverEndereco(Integer clienteId, Integer enderecoId) {
        if (enderecoId != null && endereco.buscarEnderecoPorId(clienteId, enderecoId) != null) {
            return enderecoId;
        }

        var primeiroEndereco = endereco.buscarPrimeiroEndereco(clienteId);
        if (primeiroEndereco != null) {
            return ((Number) primeiroEndereco.get("id")).intValue();
        }

        return endereco.criarEnderecoPadrao(clienteId);
    }

    private Integer criarPedidoComLista(Integer clienteId, Integer enderecoId, List<Map<String, Object>> itens) {
        double total = 0;

        for (var item : itens) {
            double preco = Double.parseDouble(String.valueOf(item.getOrDefault("preco", 0)));
            int quantidade = Integer.parseInt(String.valueOf(item.getOrDefault("quantidade", 1)));
            total += preco * quantidade;
        }

        total += 8;

        Integer id = pedido.criarPedido(clienteId, enderecoId, total);
        pedido.inserirHistorico(id, "REGISTRADO");
        pedido.inserirItensPedido(id, itens);

        return id;
    }

    public List<Map<String, Object>> listarPedidos(Integer clienteId) {
        List<Map<String, Object>> pedidos = pedido.listarPedidosPorCliente(clienteId);

        for (Map<String, Object> p : pedidos) {
            Integer pedidoId = (Integer) p.get("id");

            List<Map<String, Object>> itens = pedido.listarItensPedido(pedidoId);

            p.put("itens", itens);
        }

        return pedidos;
    }

    
    public Map<String, Object> detalharPedido(Integer clienteId, Integer pedidoId) {
        var p = pedido.buscarPedidoPorId(pedidoId, clienteId);
        if (p == null)
            throw new RuntimeException("Pedido não encontrado.");

        p.put("itens", pedido.listarItensPedido(pedidoId));

        return p;
    }

    public Map<String, Object> detalharPedidoAdmin(Integer pedidoId) {
        var p = pedido.buscarPedidoPorIdSemCliente(pedidoId);
        if (p == null)
            throw new RuntimeException("Pedido não encontrado.");

        p.put("itens", pedido.listarItensPedido(pedidoId));

        return p;
    }

    public Map<String, Object> buscarStatusPedido(Integer pedidoId) {
        var p = pedido.buscarStatusPedido(pedidoId);
        if (p == null)
            throw new RuntimeException("Pedido não encontrado.");

        return p;
    }

    @Transactional
    public void atualizarStatusPedido(Integer id, String status) {
        status = String.valueOf(status)
                .trim()
                .toUpperCase()
                .replace(" ", "_")
                .replace("-", "_");

        if (!List.of("REGISTRADO", "PREPARANDO", "EM_ROTA", "ENTREGUE", "FINALIZADO").contains(status))
            throw new RuntimeException("Status inválido.");

        if (pedido.atualizarStatus(id, status) == 0)
            throw new RuntimeException("Pedido não encontrado.");

        pedido.inserirHistorico(id, status);
    }

    public List<Map<String, Object>> listarPedidosAdmin() {
        List<Map<String, Object>> pedidos = pedido.listarPedidosAdmin();

        for (Map<String, Object> p : pedidos) {
            Integer pedidoId = ((Number) p.get("id")).intValue();
            p.put("itens", pedido.listarItensPedido(pedidoId));
        }

        return pedidos;
    }
}

File: Projeto Didio Pizza/Didio Pizza Node/backend/src/service/pedidoService.js
import * as carrinhoRepository from '../repository/carrinhoRepository.js';
import * as pedidoRepository from '../repository/pedidoRepository.js';
import * as enderecoRepository from '../repository/enderecoRepository.js';
import * as lojaService from './lojaService.js';
import connection from '../repository/connection.js';


export async function fecharPedido(clienteId, enderecoId) {

    await lojaService.verificarLojaAberta();

    const conn = await connection.getConnection();

    try {
        await conn.beginTransaction();

        let endereco = await enderecoRepository.buscarEnderecoPorId(clienteId, enderecoId);

        if (!endereco) {
            throw new Error('Endereço não encontrado.');
        }

        let carrinho = await carrinhoRepository.buscarCarrinhoAtivo(clienteId);

        if (!carrinho) {
            throw new Error('Carrinho vazio.');
        }

        let itens = await carrinhoRepository.listarItens(carrinho.id);

        if (itens.length == 0) {
            throw new Error('Carrinho sem itens.');
        }

        let total = 0;

        for (let item of itens) {
            total += item.preco * item.quantidade;
        }

        let pedidoId = await pedidoRepository.criarPedido(conn, clienteId, enderecoId, total);

        await pedidoRepository.inserirHistorico(conn, pedidoId, 'REGISTRADO');

        await pedidoRepository.inserirItensPedido(conn, pedidoId, itens);

        await pedidoRepository.limparCarrinho(conn, carrinho.id);

        await conn.commit();

        return pedidoId;

    } catch (err) {
        await conn.rollback();
        throw err;
    } finally {
        conn.release();
    }
}


export async function listarPedidos(clienteId) {

    let pedidos = await pedidoRepository.listarPedidosPorCliente(clienteId);
    return pedidos;
}


export async function detalharPedido(clienteId, pedidoId) {

    let pedido = await pedidoRepository.buscarPedidoPorId(pedidoId, clienteId);

    if (!pedido) {
        throw new Error('Pedido não encontrado.');
    }

    let itens = await pedidoRepository.listarItensPedido(pedidoId);

    pedido.itens = itens;

    return pedido;
}


export async function atualizarStatusPedido(pedidoId, status) {

    const statusValidos = ['REGISTRADO', 'PREPARANDO', 'EM_ROTA', 'ENTREGUE'];
    status = String(status).toUpperCase();

    if (!statusValidos.includes(status)) {
        throw new Error('Status inválido.');
    }

    const conn = await connection.getConnection();

    try {
        await conn.beginTransaction();

        let linhas = await pedidoRepository.atualizarStatus(conn, pedidoId, status);

        if (linhas === 0) {
            throw new Error('Pedido não encontrado.');
        }

        await pedidoRepository.inserirHistorico(conn, pedidoId, status);

        await conn.commit();

    } catch (err) {
        await conn.rollback();
        throw err;
    } finally {
        conn.release();
    }
}


export async function listarPedidosAdmin() {

    let pedidos = await pedidoRepository.listarPedidosAdmin();

    return pedidos;
}